File: ruby/testcode/ssrf_045.rb
require_relative 'shared'

module WickedPdfClient
  def self.new; self; end
  def self.pdf_from_url(url); "pdf:#{url}"; end
end

REPORTS_BASE = 'https://reports.internal.example.com'.freeze

# vuln-code-snippet start ruby_ssrf_pdf_fixed_url
def generate_report_pdf(req)
  report_id = Integer(req.param('report_id'))
  WickedPdfClient.new.pdf_from_url("#{REPORTS_BASE}/#{report_id}") # vuln-code-snippet safe-line ruby_ssrf_pdf_fixed_url
  BenchmarkResponse.json({ ok: true })
end
# vuln-code-snippet end ruby_ssrf_pdf_fixed_url

File: php/testcode/ssrf_045.php
<?php
require_once __DIR__ . '/shared.php';

// vuln-code-snippet start php_ssrf_inet_pton_range_check
function ssrf045(BenchmarkRequest $req): BenchmarkResponse {
    $host = $req->param('host');
    $bin = inet_pton($host);
    if ($bin === false || (ord($bin[0]) & 0xFE) === 0xFE) {
        return BenchmarkResponse::badRequest('invalid');
    }
    $content = file_get_contents('http://' . $host . '/data'); // vuln-code-snippet safe-line php_ssrf_inet_pton_range_check
    return BenchmarkResponse::ok($content);
}
// vuln-code-snippet end php_ssrf_inet_pton_range_check
